### ISSUE TITLE
Azure Functions cannot discover 'MAIN_CLASS'

### ISSUE THREAD

**Issue by miller79** [spring-cloud/spring-cloud-function#431]
# Version Used
- Spring Boot 2.2.2.RELEASE
- Spring Cloud Hoxton.RELEASE

# Problem Description
After upgrading from Spring Boot 2.1.X and Spring Cloud Greenwich.X, I am receiving the following error when running an Azure Function:

```
[12/11/2019 5:29:51 PM] System.Private.CoreLib: Exception while executing function: Functions.casCleanup. System.Private.CoreLib: Result: Failure
Exception: IllegalArgumentException: Failed to locate main class
Stack: java.lang.IllegalStateException: Failed to discover main class. An attempt was made to discover main class as 'MAIN_CLASS' environment variable, system property as well as entry in META-INF/MANIFEST.MF (in that order).
[12/11/2019 5:29:51 PM] 	at org.springframework.cloud.function.context.AbstractSpringFunctionAdapterInitializer.getStartClass(AbstractSpringFunctionAdapterInitializer.java:288)
[12/11/2019 5:29:51 PM] 	at org.springframework.cloud.function.context.AbstractSpringFunctionAdapterInitializer.<init>(AbstractSpringFunctionAdapterInitializer.java:100)
[12/11/2019 5:29:51 PM] 	at org.springframework.cloud.function.adapter.azure.AzureSpringBootRequestHandler.<init>(AzureSpringBootRequestHandler.java:44)
```

Looking in the built jar file, I am seeing the following:

```
Manifest-Version: 1.0
Implementation-Title: sample-project
Implementation-Version: LOCAL-SNAPSHOT
Start-Class: sample.Application
Spring-Boot-Classes: 
Build-Jdk-Spec: 1.8
Spring-Boot-Version: 2.2.2.RELEASE
Created-By: Maven Archiver 3.4.0
Main-Class: org.springframework.boot.loader.wrapper.ThinJarWrapper
```

This is easily corrected when adding my `@SpringBootApplication` class (which is the `START-CLASS` in my manifest) to the `local.settings.json` file using the value `MAIN_CLASS`:

_Working_
```json
{
  "IsEncrypted": false,
  "Values": {
    "FUNCTIONS_EXTENSION_VERSION": "~2",
    "FUNCTIONS_WORKER_RUNTIME": "java",
    "WEBSITE_NODE_DEFAULT_VERSION": "10.1.4",
    "WEBSITE_RUN_FROM_PACKAGE": "1",
    "SPRING_PROFILES_ACTIVE": "LOCAL,DEV",
    "MAIN_CLASS": "sample.Application"
  },
  "ConnectionStrings": {}
}
```

_NonWorking_
```json
{
  "IsEncrypted": false,
  "Values": {
    "FUNCTIONS_EXTENSION_VERSION": "~2",
    "FUNCTIONS_WORKER_RUNTIME": "java",
    "WEBSITE_NODE_DEFAULT_VERSION": "10.1.4",
    "WEBSITE_RUN_FROM_PACKAGE": "1",
    "SPRING_PROFILES_ACTIVE": "LOCAL,DEV"
  },
  "ConnectionStrings": {}
}
```

Looking at the `AbstractSpringFunctionAdapterInitializer`, it seems that it is looking specifically for MAIN_CLASS and I'm wondering if it should be changed to look for START_CLASS instead to match how the Spring Boot jar file is being built.
I'm not sure if this is the issue per say but it seems to be related.

# Samples

Here is a parent pom I created to help assist with creating azure functions for teams that makes setup a bit easier:

```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.2.2.RELEASE</version>
	</parent>

	<groupId>sample</groupId>
	<artifactId>azure_functions_parent</artifactId>
	<version>LOCAL-SNAPSHOT</version>
	<packaging>pom</packaging>

	<properties>
		<java.version>1.8</java.version>
		<stagingDirectory>${project.build.directory}/azure-functions/${functionAppName}</stagingDirectory>
		<azure.functions.maven.plugin.version>1.3.4</azure.functions.maven.plugin.version>
		<azure.functions.java.library.version>1.3.0</azure.functions.java.library.version>
		<wrapper.version>1.0.23.RELEASE</wrapper.version>
	</properties>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework.cloud</groupId>
				<artifactId>spring-cloud-dependencies</artifactId>
				<version>Hoxton.RELEASE</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
			<dependency>
				<groupId>com.microsoft.azure.functions</groupId>
				<artifactId>azure-functions-java-library</artifactId>
				<version>${azure.functions.java.library.version}</version>
			</dependency>
		</dependencies>
	</dependencyManagement>

	<build>
		<pluginManagement>
			<plugins>
				<plugin>
					<groupId>org.springframework.boot</groupId>
					<artifactId>spring-boot-maven-plugin</artifactId>
					<dependencies>
						<dependency>
							<groupId>org.springframework.boot.experimental</groupId>
							<artifactId>spring-boot-thin-layout</artifactId>
							<version>${wrapper.version}</version>
						</dependency>
					</dependencies>
				</plugin>
				<plugin>
					<groupId>org.apache.maven.plugins</groupId>
					<artifactId>maven-deploy-plugin</artifactId>
					<configuration>
						<skip>true</skip>
					</configuration>
				</plugin>
				<plugin>
					<groupId>com.microsoft.azure</groupId>
					<artifactId>azure-functions-maven-plugin</artifactId>
					<version>${azure.functions.maven.plugin.version}</version>
					<configuration>
						<resourceGroup>${functionResourceGroup}</resourceGroup>
						<appName>${functionAppName}</appName>
						<region>${functionAppRegion}</region>
						<appSettings>
							<!-- Run Azure Function from package file by default -->
							<property>
								<name>WEBSITE_RUN_FROM_PACKAGE</name>
								<value>1</value>
							</property>
							<property>
								<name>FUNCTIONS_EXTENSION_VERSION</name>
								<value>~2</value>
							</property>
							<property>
								<name>FUNCTIONS_WORKER_RUNTIME</name>
								<value>java</value>
							</property>
						</appSettings>
					</configuration>
					<executions>
						<execution>
							<id>package-functions</id>
							<goals>
								<goal>package</goal>
							</goals>
						</execution>
					</executions>
				</plugin>
				<plugin>
					<groupId>org.apache.maven.plugins</groupId>
					<artifactId>maven-resources-plugin</artifactId>
					<executions>
						<execution>
							<id>copy-resources</id>
							<phase>package</phase>
							<goals>
								<goal>copy-resources</goal>
							</goals>
							<configuration>
								<overwrite>true</overwrite>
								<outputDirectory>
									${stagingDirectory}
								</outputDirectory>
								<resources>
									<resource>
										<directory>${project.basedir}/src/main/azure
										</directory>
										<includes>
											<include>**</include>
										</includes>
									</resource>
								</resources>
							</configuration>
						</execution>
					</executions>
				</plugin>
				<plugin>
					<groupId>org.apache.maven.plugins</groupId>
					<artifactId>maven-dependency-plugin</artifactId>
					<executions>
						<execution>
							<id>copy-dependencies</id>
							<phase>prepare-package</phase>
							<goals>
								<goal>copy-dependencies</goal>
							</goals>
							<configuration>
								<outputDirectory>${stagingDirectory}/lib</outputDirectory>
								<overWriteReleases>false</overWriteReleases>
								<overWriteSnapshots>false</overWriteSnapshots>
								<overWriteIfNewer>true</overWriteIfNewer>
								<includeScope>runtime</includeScope>
								<excludeArtifactIds>azure-functions-java-library</excludeArtifactIds>
							</configuration>
						</execution>
					</executions>
				</plugin>

				<!--Remove obj folder generated by .NET SDK in maven clean -->
				<plugin>
					<artifactId>maven-clean-plugin</artifactId>
					<configuration>
						<filesets>
							<fileset>
								<directory>${basedir}</directory>
								<includes>
									<include>obj/**</include>
								</includes>
							</fileset>
							<fileset>
								<directory>${basedir}</directory>
								<includes>
									<include>extensions.csproj</include>
								</includes>
							</fileset>
						</filesets>
					</configuration>
				</plugin>
			</plugins>
		</pluginManagement>
	</build>
</project>
```

Here is a sample projects pom file that utilizes this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<parent>
		<groupId>sample</groupId>
		<artifactId>azure_functions_parent</artifactId>
		<version>LOCAL-SNAPSHOT</version>
	</parent>

	<artifactId>sample-project</artifactId>
	<packaging>jar</packaging>

	<properties>
		<functionAppName>app-name</functionAppName>
		<functionAppRegion>southcentralus</functionAppRegion>
		<functionResourceGroup>myresourcegroup</functionResourceGroup>
		<start-class>sample.Application</start-class>
	</properties>

	<dependencies>
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-function-adapter-azure</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-configuration-processor</artifactId>
			<optional>true</optional>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
		<dependency>
			<groupId>org.projectlombok</groupId>
			<artifactId>lombok</artifactId>
			<optional>true</optional>
		</dependency>
	</dependencies>

	<build>
		<resources>
			<resource>
				<directory>src/main/resources</directory>
				<filtering>true</filtering>
			</resource>
		</resources>
		<plugins>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-deploy-plugin</artifactId>
			</plugin>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-dependency-plugin</artifactId>
			</plugin>
			<plugin>
				<groupId>com.microsoft.azure</groupId>
				<artifactId>azure-functions-maven-plugin</artifactId>
			</plugin>
			<plugin>
				<artifactId>maven-resources-plugin</artifactId>
			</plugin>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		</plugins>
	</build>
</project>
```

I've created the following classes for a simple test:

```java
package sample;

import java.util.function.Function;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;

import com.microsoft.azure.functions.ExecutionContext;

import reactor.core.publisher.Flux;

@SpringBootApplication
public class Application {
	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

	@Bean
	public Function<Flux<String>, Flux<String>> uppercase(ExecutionContext context) {
		return flux -> flux.map(value -> value.toUpperCase());
	}
}
```

```java
package sample;

import java.util.Optional;

import org.springframework.cloud.function.adapter.azure.AzureSpringBootRequestHandler;

import com.microsoft.azure.functions.ExecutionContext;
import com.microsoft.azure.functions.HttpMethod;
import com.microsoft.azure.functions.HttpRequestMessage;
import com.microsoft.azure.functions.HttpResponseMessage;
import com.microsoft.azure.functions.HttpStatus;
import com.microsoft.azure.functions.annotation.AuthorizationLevel;
import com.microsoft.azure.functions.annotation.FunctionName;
import com.microsoft.azure.functions.annotation.HttpTrigger;

public class Handler extends AzureSpringBootRequestHandler<String, String> {
	@FunctionName("uppercase")
	public HttpResponseMessage execute(@HttpTrigger(name = "req", methods = { HttpMethod.GET,
			HttpMethod.POST }, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<Optional<String>> request,
			ExecutionContext context) {
		// Parse query parameter
		String query = request.getQueryParameters().get("value");
		String name = request.getBody().orElse(query);

		if (name == null) {
			return request.createResponseBuilder(HttpStatus.BAD_REQUEST).body("Please pass a value").build();
		} else {
			return request.createResponseBuilder(HttpStatus.OK).body(handleRequest(name, context)).build();
		}
	}
}
```

**Comment by olegz** [comment]
@miller79 this issue has literally just been fixed few hours ago. Can you please try with spring-cloud-function 3.0.1.BUILD-SNAPSHOT and let us know?

**Comment by miller79** [comment]
Just tried it with the following and I'm still receiving the same error:

- spring-cloud-function-adapter-azure: 3.0.1.BUILD-20191211.181256-45
- spring-cloud-function-context: 3.0.1.BUILD-20191211.180745-45
- spring-cloud-function-core: 3.0.1.BUILD-20191211.180639-45

```
```

**Comment by olegz** [comment]
are you on https://gitter.im/spring-cloud/spring-cloud-stream gitter by any chance?

**Comment by miller79** [comment]
I haven't been in a while but I can sign in.

**Comment by olegz** [comment]
Actually never mind, let me look into this a bit deeper and i'll get back to you

**Comment by olegz** [comment]
I was able to reproduce it, so the fix will be in 3.0.1.

**Comment by olegz** [comment]
Also, this is related to https://github.com/spring-cloud/spring-cloud-function/issues/421
